seastar/src/submit_to.rs

```diff
@@ -0,0 +1,122 @@
+            closure: *mut u8,
+            caller: unsafe fn(*mut u8) -> VoidFuture,
+        ) -> VoidFuture;
+    }
+}
+
+/// Convert the pointer to Box<Func> and then call it, consuming it in the process.
+fn fn_once_caller<Func>(raw_func: *mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    unsafe { Box::from_raw(raw_func as *mut Func)() }
+}
+
+/// A helper function. We need to be able to name the type of a named closure.
+/// There is no `typeof` in Rust, so this is achieved by using a dummy paramter.
+const fn get_fn_once_caller<Func>(_: &Func) -> fn(*mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    fn_once_caller::<Func>
+}
+
+/// Free a pointer. We need to call this function in the case of an error.
+fn fn_once_dropper<Func>(raw_func: *mut u8) {
+    unsafe {
+        let _ = Box::from_raw(raw_func as *mut Func);
+    }
+}
+
+const fn get_fn_once_dropper<Func>(_: &Func) -> fn(*mut u8) {
+    fn_once_dropper::<Func>
+}
+
+/// Runs a function `func` on a `shard_id` shard.
+///
+/// # Example
+///
+/// ```rust
+/// #[seastar::test]
+/// async fn submit_to_example() {
+///     let ret = submit_to(0, || future::ready(42)).await;
+///     assert!(matches!(ret, Ok(42)));
+/// }
+pub async fn submit_to<Func, Fut, Ret>(shard_id: u32, func: Func) -> cxx_async::CxxAsyncResult<Ret>
+where
+    Func: FnOnce() -> Fut + Send + 'static,
+    Fut: Future<Output = Ret> + 'static,
+    Ret: Send + 'static,
+{
+    let (tx, rx) = futures::channel::oneshot::channel::<Ret>();
+
+    let closure = move || {
+        VoidFuture::infallible_local(async {
+            tx.send(func().await).ok();
+        })
+    };
+
+    let closure_caller = get_fn_once_caller(&closure);
+    let dropper = get_fn_once_dropper(&closure);
+    let boxed_closure = Box::into_raw(Box::new(closure)) as *mut u8;
+
+    unsafe {
+        match ffi::submit_to(shard_id, boxed_closure, closure_caller).await {
+            Ok(_) => Ok(rx.await.unwrap()),
+            Err(e) => {
+                dropper(boxed_closure);
+                Err(e)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[seastar::test]
+    async fn test_submit_to() {
+        let ret = submit_to(0, || async { 42 }).await;
+        assert!(matches!(ret, Ok(42)));
+    }
+
+    #[seastar::test]
+    async fn test_submit_to_nested() {
+        let ret = submit_to(0, || async { submit_to(0, || async { 42 }).await.unwrap() }).await;
+        assert!(matches!(ret, Ok(42)));
+    }
+
+    #[seastar::test]
+    async fn test_submit_to_two_shards() {
+        let ret = submit_to(0, || async { 17 }).await;
+        assert!(matches!(ret, Ok(17)));
+        let ret = submit_to(1, || async { 25 }).await;
+        assert!(matches!(ret, Ok(25)));
+    }
+
+    #[seastar::test]
+    async fn test_submit_to_two_shards_nested() {
+        let ret = submit_to(0, || async { submit_to(1, || async { 42 }).await.unwrap() }).await;
```

Comment:
Nit: it would be slightly more interesting to submit the outer task to shard 1 and the inner to 0 - so that both submits need to cross the shard boundary.